test/offline/test_cql.py

```diff
@@ -68,15 +68,43 @@ def get_args():
     return parser.parse_known_args()[0]
 
 
-def test_cql(args=get_args()):
```

Comment:
From now on, more unit tests should be added to tianshou - just relying on integration tests can lead to serious bugs. Pls make sure to add tests that verify that the extension is doing what it's supposed to do.

Reply:
Yes I agree, however not sure how can we test if an RL algorithm is really doing what it is intending to do. One way would be to be to save some deterministic values that have been identified as the values that should be reproduced if everything is working ok, and then reproduce them in test. However, this is practically impossible given that any change in the software stack can affect those values and trigger a false negative. 
Do you have some suggestions on how to do it?

Reply:
Testing performance is indeed difficult, but unit testing should be ok. The fact that it's difficult now is a reflection of imperfect design, like the very long implementation of `learn`. If you create a new method that returns what is now called `catq1/2`, you could test then when `calibrated=True`, the max is <= the max of the returns.

You can also test that the processed buffer has the correct field if `calibrated=True` after preprocess of the buffer, as discussed above.

Adding a second integration test with `calibration=True` serves relatively little purpose, but increases the runtime by a lot. Over time, the tests will be adjusted to be more useful and faster - this will require breaking up long methods in smaller ones. It's a lot of work, but it's very important for the overall future of tianshou and the developer experiences. Let's start this process now, ok? At least, new code shouldn't add to the current problems of tianshou :) 

Thank you again for your contribution and patience, I appreciate it!

Reply:
Regarding performance - monkey patching is the way to go here, it allows mimicking performance outputs but testing everything else. But that goes beyond the scope of this PR

Reply:
Thanks for your thorough reviews and your contributions as well :)
For the testing, I think what you are proposing boils down to testing if the max operation is done correctly and if the returns were calculated correctly.
I don't think testing the correctness of RL algorithm implementations can be automated in this way (or at all unless more sophisticated AI is invented).

Reply:
I tested those manually to check if the implementation is correct, and I don't think it is useful to test it every time. Until a better system is developed I don't run that running tests that verify those two things makes sense when all the other parts of the algorithm are not tested and as I mentioned I believe that it's not something that can be automatized. 
I think you are right about avoiding to run CQL two times, I'll set calibrated to True as default and just run CQL once. In my testing the calibrated version performs better across the board so it might as well be the default.

Reply:
It's not only about being correct now but also about future developers who want to know that they haven't broken stuff by making changes. The latter is very important. 

Apart from that, the absence of unit tests in tianshou has led to untestable (i.e. not will designed) code, as you have encountered now. At least I would want new contributions to make the overall code quality slightly better, not slightly worse.

But I understand if you don't want to invest additional time on it now. We can merge this first and I will add a unit test for this in a follow up, and assign you as reviewer, if you prefer that

Reply:
For me it's just in principle it doesn't make sense to have automated tests for RL algos otherwise we should test every line of the code which doesn't make sense. I believe RL algos should be tested manually by looking at losses and other metrics not this way. If I'm convinced otherwise I would happily do it.

Reply:
I don't think I can say something convincing here, so let me do a follow-up PR. Maybe the way I write it will give you a better feeling for what I mean